auto_dev/protocols/formatter.py

```diff
@@ -0,0 +1,247 @@
+        case ast.OneOf:
+            if not all(isinstance(e, Field) for e in element.elements):
+                msg = "Only implemented OneOf for Field"
+                raise NotImplementedError(msg)
+            inner = " | ".join(render_field(e, message) for e in element.elements)
+            return f"{element.name}: {inner}"
+        case adapters.MessageAdapter:
+            elements = sorted(element.elements, key=lambda e: not isinstance(e, MessageAdapter | ast.Enum))
+            body = inner = "\n".join(render_attribute(e, element) for e in elements)
+            encoder = render_encoder(element)
+            decoder = render_decoder(element)
+            body = f"{inner}\n\n{encoder}\n\n{decoder}"
+            indented_body = textwrap.indent(body, "    ")
+            return f"\nclass {element.name}(BaseModel):\n" f'    """{element.name}"""\n\n' f"{indented_body}\n"
+        case ast.Enum:
+            members = "\n".join(f"{val.name} = {val.number}" for val in element.elements)
+            indented_members = textwrap.indent(members, "    ")
+            return f"class {element.name}(IntEnum):\n" f'    """{element.name}"""\n\n' f"{indented_members}\n"
+        case ast.MapField:
+            key_type = PRIMITIVE_TYPE_MAP.get(element.key_type, element.key_type)
+            value_type = qualified_type(message, element.value_type)
+            return f"{element.name}: dict[{key_type}, {value_type}]"
+        case ast.Group | ast.Option | ast.ExtensionRange | ast.Reserved | ast.Extension:
+            msg = f"{element}"
+            raise NotImplementedError(msg)
+        case _:
+            msg = f"Unexpected message type: {element}"
+            raise TypeError(msg)
+
+
+def render(file: FileAdapter):
+    """Main function to render a .proto file."""
+
+    enums = "\n".join(render_attribute(e, file) for e in file.enums)
+    messages = "\n".join(render_attribute(e, file) for e in file.messages)
+
+    return f"{enums}\n{messages}"
+
+
+def encode_field(element, message):
+    """Render pydantic model field encoding."""
+
+    instance_attr = f"{message.name.lower()}.{element.name}"
+    if (
+        element.type in PRIMITIVE_TYPE_MAP
+        or element.type in message.enum_names
+        or element.type in message.file.enum_names
+    ):
+        value = instance_attr
+    else:  # Message
+        qualified = qualified_type(message, element.type)
+        if element.cardinality == FieldCardinality.REPEATED:
+            return f"for item in {instance_attr}:\n" f"    {qualified}.encode(proto_obj.{element.name}.add(), item)"
+        if element.cardinality == FieldCardinality.OPTIONAL:
+            return (
+                f"if {instance_attr} is not None:\n"
+                f"    temp = proto_obj.{element.name}.__class__()\n"
+                f"    {qualified}.encode(temp, {instance_attr})\n"
+                f"    proto_obj.{element.name}.CopyFrom(temp)"
+            )
```

Comment:
This was somewhat interesting. Unlike with optional `Fields`, where the attribute won't be present when not set (`None` in python) and where one has to use `proto_obj.HasField()` to check if it exists, a nested optional `Message` is always present when the parent is instantiated. At the same time, `proto_obj.HasField()` will return `False` despite the default instantiation.

We need to differentiate between a empty `Message` and a null value. Since the `.encode` on the pydantic model for an empty message does nothing and returns `None`, we need to instantiate it prior to calling the nested models `.decode()` to be able to deal correctly with both empty and non-empty messages.